hail/python/hailtop/tls.py

```diff
@@ -61,6 +62,22 @@ def get_in_cluster_client_ssl_context() -> ssl.SSLContext:
     return client_ssl_context
 
 
+def get_in_cluster_no_hostname_checks_client_ssl_context() -> ssl.SSLContext:
+    global no_hostname_checks_client_ssl_context
+    if no_hostname_checks_client_ssl_context is None:
+        ssl_config = _get_ssl_config()
+        no_hostname_checks_client_ssl_context = ssl.create_default_context(
+            purpose=Purpose.SERVER_AUTH,
+            cafile=ssl_config['outgoing_trust'])
+        no_hostname_checks_client_ssl_context.load_cert_chain(
+            ssl_config['cert'],
+            keyfile=ssl_config['key'],
```

Comment:
For my education, can you give me the high level view of why you needed to add this function for the address service only?

Reply:
Line 77 is the key line. It sets check_hostname to `False`.

In Figure 11-5 on page 137 of Rice's "Container Security", we see that after the SYN/ACK and HELLOs, the server sends its certificate to the client. The client found this server directly with an IP address or indirectly by looking up a domain name. Every certificate has a list of acceptable "names" for servers using this certificate. Those "names" can be IP addresses or domain names. If `check_hostname` is set to `True`, then the client will terminate the *TLS handshake* if the name the client used does not appear in the acceptable names list.

The address service is replacing DNS. Normally DNS is performed by the TLS library itself. We can't replace DNS directly. Instead, we resolve the k8s service name first, using `DeployConfig.address`. Then we pass an IP address to the TLS library. The TLS library looks for the IP address in the acceptable names. It will not find the name and thus terminate the handshake in failure.

We cannot place pod IP addresses inside the certificates because a pod could die and return with a new IP address. We would have to change certificates every time a pod died. That's not tractable.

The long term solution is to replace DNS directly. We could modify /etc/hosts on every machine. We could try and use Google CloudDNS directly. We could use a mixture of DNS and iptables. We could literally implement the DNS protocol and point all our VMs at our new DNS service.

Reply:
See test_address.py for an example of looking up the IP address first then making an HTTPS request using the IP address.

Reply:
You got rid of the check_hostname to False argument. Is this still needed then?